File: c++/079.cpp
class Solution {
    bool** visit;
    vector<vector<char>> bd;
    string w;
    int m,n;
    bool res=false;
public:
    bool exist(vector<vector<char>>& board, string word) {
        this->bd.assign(board.begin(),board.end());
        this->w = word;
        m=board.size(), n=board[0].size();
        visit = new bool*[m];
        for(int i=0; i<m; i++){
            visit[i] = new bool[n];
            for(int j=0; j<n; j++)
                visit[i][j]=0;
        }
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(!res&&board[i][j]==word[0]){
                    dfs(i,j,1);
                    visit[i][j]=0;
                }
            }
        }
        for(int i=0; i<m; i++)
            delete[] visit[i];
        delete[] visit;
        return res;
    }
    void dfs(int a, int b, int k){
        if(k>=w.length()||res){
            res = true;
            return;
        }
        visit[a][b]=1;
        int ar[4]={a-1,a+1,b-1,b+1};
        for(int i=0; i<4; i++){
            if(i<2){
                if(ar[i]>=0&&ar[i]<m&&!visit[ar[i]][b]&&bd[ar[i]][b]==w[k]){
                    dfs(ar[i],b,k+1);
                    visit[ar[i]][b]=0;
                }
            }else{
                if(ar[i]>=0&&ar[i]<n&&!visit[a][ar[i]]&&bd[a][ar[i]]==w[k]){
                    dfs(a,ar[i],k+1);
                    visit[a][ar[i]]=0;
                }
            }
        }
    }
};

File: go/079.go
package main

/**
单词搜索
二位字符网格中寻找字符串  连续 不重复
思路：dfs 回溯
*/
func exist(board [][]byte, word string) bool {
	visited := make([][]bool, len(board))
	for i := 0; i < len(visited); i++ {
		visited[i] = make([]bool, len(board[0]))
	}
	for i, v := range board {
		for j := range v {
			if d(board, visited, word, i, j, 0) {
				return true
			}
		}
	}
	return false
}

func d(board [][]byte, visited [][]bool, word string, x, y, idx int) bool {
	n, m := len(board), len(board[0])
	if idx == len(word)-1 {
		return board[x][y] == word[idx]
	}
	// 上右下左搜索
	dx, dy := []int{-1, 0, 1, 0}, []int{0, 1, 0, -1}
	if board[x][y] == word[idx] {
		visited[x][y] = true
		for i := 0; i < 4; i++ {
			ix, iy := x+dx[i], y+dy[i]
			f := ix >= 0 && ix < n && iy >= 0 && iy < m
			if f && d(board, visited, word, ix, iy, idx+1) {
				return true
			}
		}
		visited[x][y] = false
	}
	return false
}
